air/src/utils.rs

```diff
@@ -0,0 +1,216 @@
+pub fn checked_andn<F: Field>(x: F, y: F) -> F {
+    debug_assert!(x.is_zero() || x.is_one());
+    debug_assert!(y.is_zero() || y.is_one());
+    andn(x, y)
+}
+
+/// Convert a 32-bit integer into an array of 32 0 or 1 field elements.
+pub fn u32_to_bits<AF: AbstractField>(val: u32) -> [AF; 32] {
+    // We do this over F::from_canonical_u32 as from_canonical_u32 can be slow
+    // like in the case of monty field.
+    array::from_fn(|i| {
+        if val & (1 << i) != 0 {
+            AF::ONE
+        } else {
+            AF::ZERO
+        }
+    })
+}
+
+/// Compute a + b mod 2^32 returning the summation and data allowing us to verify the summation.
+///
+/// The data we need is whether a + b > 2^32 and similarly
+/// whether (a mod 2^16) + (b mod 2^16) > 2^16;
+pub fn verifiable_add(a: u32, b: u32) -> (u32, [bool; 2]) {
+    let a_16 = a as u16;
+    let b_16 = b as u16;
+
+    let (_, overflow_16) = a_16.overflowing_add(b_16);
+    let (sum, overflow) = a.overflowing_add(b);
+
+    (sum, [overflow, overflow_16])
+}
+
+/// Verify that `a = b + c + d mod 2^32`
+///
+/// We assume that a, b, c, d are all given as `2 16` bit limbs (e.g. `a = a[0] + 2^16a[1]` which
+/// have been range checked separately. This verifies that the auxiliary variables lie in {0, 1, 2} and satisfy
+///
+/// `2^32 aux[0] + a = b + c + d mod P`
+///
+/// `2^16 aux[1] + a[0] = b[0] + c[0] + d[0] mod P`
+pub fn triple_add<AB: AirBuilder>(
+    builder: &mut AB,
+    a: &[<AB as AirBuilder>::Var; 2],
+    b: &[<AB as AirBuilder>::Var; 2],
+    c: &[<AB as AirBuilder>::Expr; 2],
+    d: &[<AB as AirBuilder>::Expr; 2],
+    aux: &[<AB as AirBuilder>::Var; 2],
+) {
+    // We add four constraints:
+    //
+    // aux[0] * (aux[0] - 1) * (aux[0] - 2) = 0
+    // aux[1] * (aux[1] - 1) * (aux[1] - 2) = 0
+    // 2^32 aux[0] + a - b - c - d = 0
+    // 2^16 aux[1] + a[0] - b[0] - c[0] - d[0] = 0
+    //
+    // The first two check that both auxiliary variables lie in {0, 1, 2}.
+    // The third equation checks that the sum `2^32 aux[0] + a = b + c + d` holds mod P.
+    // For the final equation, note that overflow is impossible and `x[0] = x mod 2^16`.
+    // Thus if the final equation holds then `2^32 aux[0] + a = b + c + d mod 2^{16}`
+    //
+    // Using the chinese remainder theorem we conclude `a = b + c + d mod 2^32`
+
+    builder.assert_tern(aux[0]);
+    builder.assert_tern(aux[1]);
+
+    // TODO: Ideally these should eventually be saved as constants. Or at the very least
+    // there should be a quicker method of generating them than exp_u64.
+    let two_16 = <AB as AirBuilder>::Expr::TWO.exp_u64(16);
+    let two_32 = <AB as AirBuilder>::Expr::TWO.exp_u64(32);
+
+    let sum_16 = a[0] - b[0] - c[0].clone() - d[0].clone();
+    let sum_32 = a[1] - b[1] - c[1].clone() - d[1].clone();
+
+    builder.assert_zero(two_32 * aux[0] + sum_16.clone() + two_16.clone() * sum_32);
+    builder.assert_zero(two_16 * aux[1] + sum_16);
```

Comment:
Both of these constraints should be linear (as long as `c, d` are), which means we can instead compute `aux` directly from the inputs and apply the ternary constraint over this expression. This avoid the two extra auxiliary witness columns and the `assert_zero` constraints

Reply:
Nice observation, thanks for pointing it out.